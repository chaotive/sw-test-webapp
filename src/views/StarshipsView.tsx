import React from 'react'
import Container from 'react-bootstrap/Container'
import {Starship} from '../../typings/swapi/starships'
import {loadingFragment, titleFragment} from '../components/fragments'
import {retrieveStarships} from '../requests/swapi'
import StarshipCapsule from '../components/StarshipCapsule'
import {IPaginable} from '../../typings/traits'
import PagingButtons from '../components/PagingButtons'

interface IStarshipsViewState {
  loaded?: boolean
  nextPage?: string
  previousPage?: string
  starships: Starship[]
  pageNumber: number
}

function updateState(c: StarshipsView, page?: string, pageDiff: number = 0) {
  c.setState({
    nextPage: undefined,
    previousPage: undefined,
    loaded: false,
    pageNumber: c.state.pageNumber + pageDiff
  })
  retrieveStarships(page).then(data =>
    c.setState({
      starships: data.results,
      nextPage: data.next,
      previousPage: data.previous,
      loaded: true
    })
  )
}

const renderStarships = (c: StarshipsView) => c.state.starships.map((s, i)=> <StarshipCapsule key={i} {...s}/>)
const goBack = (c: StarshipsView) => updateState(c, c.state.previousPage, -1)
const goForward = (c: StarshipsView) => updateState(c,  c.state.nextPage, 1)

export default class StarshipsView extends React.Component<{}, IStarshipsViewState> implements IPaginable {
  constructor(props: {}) {
    super(props)
    this.state = {
      pageNumber: 1,
      starships: []
    }
  }

  public goBack = () => goBack(this)
  public goForward = () => goForward(this)
  public componentWillMount = () => updateState(this)

  public render() {
    return <Container>
      {titleFragment('Starships page ' + this.state.pageNumber)}
      {this.state.loaded ? renderStarships(this) : loadingFragment}
      <PagingButtons {...this} />
    </Container>
  }
}
