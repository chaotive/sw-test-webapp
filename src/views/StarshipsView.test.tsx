import React from 'react'
import ReactDOM from 'react-dom'
import StarshipsView from './StarshipsView'
import renderer from 'react-test-renderer'

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<StarshipsView/>, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders correctly', () => {
  const tree = renderer
    .create(<StarshipsView/>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
