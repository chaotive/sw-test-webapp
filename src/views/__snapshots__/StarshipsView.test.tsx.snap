// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <h1
      className="col"
    >
      STAR WARS Starships page 1
    </h1>
  </div>
  <div
    className="m-4 row"
  >
    <div
      className="col"
    >
      <button
        className="btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Back
      </button>
    </div>
    <div
      className="col"
    >
      <button
        className="float-right btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Forward
      </button>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="text-center m-4 col"
    >
      <div
        className="spinner-border"
        role="status"
      >
        <span
          className="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
